File: cdk_lib/lambda_function.py
from aws_cdk import aws_lambda, core, aws_s3, aws_s3_notifications
from aws_cdk.aws_dynamodb import Table
from aws_cdk.aws_s3 import EventType, NotificationKeyFilter
from aws_cdk.core import Stack

from cdk_lib.string_utils import snake

DEFAULT_RUNTIME = aws_lambda.Runtime.PYTHON_3_8


def lambda_function(scope: core.Construct, name: str, **kwargs) -> aws_lambda.Function:
    """Creates a :py:class:`aws_lambda.Function` with default arguments"""
    defaults = {
        'function_name': f'{name}-fn',
        'runtime': DEFAULT_RUNTIME,
        'code': scope.code_asset('src'),
        'handler': f'{snake(Stack.of(scope).stack_name)}.{snake(name)}_lambda.handler',
        'layers': getattr(scope, 'lambda_layers', []),
    }

    function_id = name.title() + 'Handler'
    return aws_lambda.Function(scope, function_id, **defaults, **kwargs)


def s3_event_handler(scope: core.Construct, name: str,
                     bucket: aws_s3.Bucket,
                     *filters: NotificationKeyFilter,
                     s3_key_prefix=None,
                     event: EventType = aws_s3.EventType.OBJECT_CREATED,
                     event_log_table: Table = None,
                     **kwargs) -> aws_lambda.Function:
    event_handler = lambda_function(scope, name, **kwargs)
    if s3_key_prefix and not filters:
        filters = [aws_s3.NotificationKeyFilter(prefix=s3_key_prefix)]
    bucket.add_event_notification(
        event,
        aws_s3_notifications.LambdaDestination(event_handler),
        *filters
    )
    bucket.grant_read(event_handler)
    if event_log_table:
        event_log_table.grant_read_write_data(event_handler)
        event_log_table.grant(event_handler, 'dynamodb:DescribeTable')
    return event_handler

File: cdk/lib/common/lambda-function.ts
import {Function, Runtime} from "@aws-cdk/aws-lambda"
import {Duration, Stack} from "@aws-cdk/core"
import {kebab, snake, title} from "./string-utils"
import {FunctionProps} from "@aws-cdk/aws-lambda/lib/function"
import {PythonStack} from "./stack"
import {Table} from "@aws-cdk/aws-dynamodb"
import {Bucket, EventType, NotificationKeyFilter} from "@aws-cdk/aws-s3"
import {LambdaDestination} from "@aws-cdk/aws-s3-notifications"

const DEFAULT_RUNTIME = Runtime.PYTHON_3_8


function createPythonLambda(stack: PythonStack, name: string, options?: FunctionProps): Function {
    const defaults: FunctionProps = {
        functionName: `${kebab(name)}-fn`,
        runtime: DEFAULT_RUNTIME,
        timeout: Duration.seconds(30),
        code: stack.createCodeAsset('src'),
        handler: `${snake(Stack.of(stack).stackName)}.${snake(name)}_lambda.handler`,
        layers: stack.lambdaLayers,
        environment: {
            LOG_LEVEL: 'INFO',
            POWERTOOLS_SERVICE_NAME: name
        }
    }

    const functionId = title(name) + 'Handler'
    return new Function(stack, functionId, {...defaults, ...options})
}

interface S3EventHandlerProps {
    bucket: Bucket,
    filters?: NotificationKeyFilter[],
    s3KeyPrefix?: string,
    event?: EventType,
    eventLogTable?: Table,
    functionProps?: FunctionProps
}

function createS3EventHandler(stack: PythonStack, name: string, props: S3EventHandlerProps): Function {
    const eventHandler = createPythonLambda(stack, name, props.functionProps)
    let {
        event = EventType.OBJECT_CREATED,
        filters = [],
        bucket,
        eventLogTable,
        s3KeyPrefix
    } = props

    if (s3KeyPrefix && !filters) {
        filters = [{prefix: props.s3KeyPrefix}]
    }

    bucket.addEventNotification(event, new LambdaDestination(eventHandler), ...filters)
    bucket.grantRead(eventHandler)
    if (eventLogTable) {
        eventLogTable.grantReadWriteData(eventHandler)
        eventLogTable.grant(eventHandler, 'dynamodb:DescribeTable')
    }
    return eventHandler
}

export {
    createPythonLambda,
    createS3EventHandler
}
